apis/projectcontour/v1/helpers.go

```diff
@@ -78,3 +83,221 @@ func (r *Route) AuthorizationContext(parent map[string]string) map[string]string
+func (dc *DetailedCondition) AddWarningf(warnType, reason, formatmsg string, args ...interface{}) {
+
+	dc.AddWarning(warnType, reason, fmt.Sprintf(formatmsg, args...))
+}
+
+// GetWarning gets an warning of the given warnType.
+// Similar to a hash lookup, will return true in the second value if a match is
+// found, and false otherwise.
+func (dc *DetailedCondition) GetWarning(warnType string) (SubCondition, bool) {
+
+	i := getIndex(warnType, dc.Warnings)
+
+	if i == -1 {
+		return SubCondition{}, false
+	}
+
+	return dc.Warnings[i], true
+}
+
+// updateReason updates a DetailedCondition's reason and message correctly
+// if they are different to the existing ones.
+// Note that this helper may be used to build the first iteration of a
+// DetailedCondition, so it's possible that `Reason` and/or `Message` will
+// be empty when it is called.
+func (dc *DetailedCondition) updateReason(reason, message string) {
+	if dc.Reason == "" {
+		dc.Reason = reason
+		dc.Message = message
+		return
+	}
+
+	if dc.Reason != reason {
+		dc.Reason = "MultipleProblems"
+		dc.Message = "Multiple problems were found, see errors or warnings for details"
+		return
+	}
+
+	// This case covers the same `Reason` being used multiple times.
+	// The only case for this is if we're adding more details about multiple errors for the same reason.
+	if dc.Message != message {
+		// Only add the message if it's not already in there somewhere.
+		if !strings.Contains(dc.Message, message) {
+			dc.Message = dc.Message + ", " + message
+		}
+	}
+}
+
+func (sc *SubCondition) updateReason(reason, message string) {
+
+	if sc.Reason == "" {
+		sc.Reason = reason
+		sc.Message = message
+		return
+	}
+
+	if sc.Reason != reason {
+		sc.Reason = "MultipleReasons"
+	}
+
+	if !strings.Contains(sc.Message, message) {
+		sc.Message = sc.Message + ", " + message
+	}
+
+}
+
+// getIndex checks if a SubCondition of type condType exists in the
+// slice, and returns its index if so. If not, returns -1.
+func getIndex(condType string, subconds []SubCondition) int {
+
+	for i, cond := range subconds {
+		if cond.Type == condType {
+			return i
+		}
+	}
+	return -1
+}
+
+// GetConditionIndex gets the index of a condition of the given type from
+// the given DetailedCondition slice, or -1 if it is not found.
+// The result of this function should only be used on a single version of a
+// DetailedCondition; the ordering of DetailedConditions is not necessarily
+// stable across trips to the apiserver.
+func GetConditionIndex(condType string, conds []DetailedCondition) int {
+
+	for i, cond := range conds {
+		if cond.Type == condType {
+			return i
+		}
+	}
+
+	return -1
+
+}
+
+// IsPositivePolarity returns true if the DetailedCondition is a positive-polarity
+// condition like `Valid` or `Ready`, and false otherwise.
+func (dc *DetailedCondition) IsPositivePolarity() bool {
+
+	switch dc.Type {
+	case "Valid":
+		return true
+	default:
+		return false
+	}
```

Comment:
Because we're going to eventually move to inlining the upstream `Condition`, we can't use a aliased type for `DetailedCondition.Type`. I've added a constant string instead.